nexus/db-queries/src/db/datastore/affinity.rs

```diff
@@ -72,6 +73,130 @@ const INSTANCE_MAX_AFFINITY_GROUPS: usize = 16;
 const INSTANCE_MAX_ANTI_AFFINITY_GROUPS: usize = 16;
 
 impl DataStore {
+    pub async fn affinity_groups_batch_lookup(
+        &self,
+        opctx: &OpContext,
+        authz_project: &authz::Project,
+        groups: &Vec<NameOrId>,
+    ) -> ListResultVec<AffinityGroupUuid> {
+        opctx.authorize(authz::Action::ListChildren, authz_project).await?;
+
+        let mut names: Vec<Name> = vec![];
+        let mut ids: Vec<Uuid> = vec![];
+        for group in groups.iter() {
+            match group {
+                NameOrId::Name(name) => names.push(name.clone().into()),
+                NameOrId::Id(id) => ids.push(*id),
+            }
+        }
+
+        use db::schema::affinity_group::dsl;
+        let result: Vec<(Uuid, Name)> = dsl::affinity_group
+            .filter(dsl::id.eq_any(ids).or(dsl::name.eq_any(names)))
+            .filter(dsl::project_id.eq(authz_project.id()))
+            .filter(dsl::time_deleted.is_null())
+            .select((dsl::id, dsl::name))
+            .get_results_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        // If a group isn't present in the result that was present in the input
+        // that means it either doesn't exist or isn't owned by the user. Either
+        // way we want to give a specific lookup error for at least the first
+        // result. It would be nice to include an aggregate error with all the
+        // missing groups.
+        for group in groups.iter() {
+            match group {
+                NameOrId::Name(name) => {
+                    if !result
+                        .iter()
+                        .any(|(_, n)| n.clone() == name.clone().into())
+                    {
+                        return Err(Error::ObjectNotFound {
+                            type_name: ResourceType::AffinityGroup,
+                            lookup_type: LookupType::ByName(name.to_string()),
+                        });
+                    }
+                }
+                NameOrId::Id(id) => {
+                    if !result.iter().any(|&(i, _)| i == *id) {
+                        return Err(Error::ObjectNotFound {
+                            type_name: ResourceType::AffinityGroup,
+                            lookup_type: LookupType::ById(*id),
+                        });
+                    }
+                }
+            }
+        }
```

Comment:
`not_found_by_id` and `not_found_by_name` might make this very slightly shorter. Also you could loop through the names and IDs vecs since you've already done the match to separate them above. Neither change is necessary though, this is all clear enough.

Reply:
Sure thing, I went ahead and made both these changes.